pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -186,6 +216,240 @@ func (r *Registry) Remove(id ipc.ID, creds *auth.Credentials) error {
 	return nil
 }
 
+// FindByID returns the queue with the specified ID and an error if the ID
+// doesn't exist.
+func (r *Registry) FindByID(id ipc.ID) (*Queue, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	mech := r.reg.FindByID(id)
+	if mech == nil {
+		return nil, syserror.EINVAL
+	}
+	return mech.(*Queue), nil
+}
+
+// Send appends a message to the message queue, and returns an error if sending
+// fails. See msgsnd(2).
+func (q *Queue) Send(ctx context.Context, m Message, b Blocker, wait bool, pid int32) error {
+	creds := auth.CredentialsFromContext(ctx)
+	if err := q.append(ctx, m, creds, pid); err != syserror.ErrWouldBlock {
+		return err
+	}
+
+	if !wait {
+		return syserror.EAGAIN
+	}
+
+	e, ch := waiter.NewChannelEntry(nil)
+	q.senders.EventRegister(&e, waiter.EventOut)
+
+	err := q.append(ctx, m, creds, pid)
+	for ; err == syserror.ErrWouldBlock; err = q.append(ctx, m, creds, pid) {
+		b.Block(ch)
```

Comment:
I think there's a fairness concern here that can manifest as a livelock. If multiple senders are competing on a congested queue, there's no guarantee a particular sender will eventually make progress.

This is ok as long as linux also doesn't guarantee fairness. There's a similar fairness issue on the recv side.

Reply:
The man pages don't mention fairness, so I suppose that it isn't guaranteed. That being said, the linux implementation uses a different approach. It loops over blocked insertions, and only wakes up a process if it will fit the queue. I'm not sure if that qualifies as guaranteeing fairness.

Reply:
Per external discussion, this is going to be addressed in a separate PR.